File: c++/leetcode/1302. Deepest Leaves Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  int deepestLeavesSum(TreeNode *root)
  {
    queue<TreeNode *> q;
    q.emplace(root);

    int ans = 0;

    while (!q.empty())
    {
      const auto sz = q.size();
      ans = 0;

      for (int i = 0; i < sz; ++i)
      {
        auto item = q.front();

        if (item->left)
          q.emplace(item->left);
        if (item->right)
          q.emplace(item->right);
        ans += item->val;
        q.pop();
      }
    }

    return ans;
  }
};

File: javascript/leetcode/1302. Deepest Leaves Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var deepestLeavesSum = function (root) {
  let ans = 0;
  let q = [root];

  while (q.length) {
    const sz = q.length;
    ans = 0;

    for (let i = 0; i < sz; ++i) {
      const front = q[0];

      if (front.left) q.push(front.left);
      if (front.right) q.push(front.right);

      ans += front.val;

      q = q.slice(1);
    }
  }

  return ans;
};
